pertpy/tools/_distance_tests.py

```diff
@@ -0,0 +1,224 @@
+        """Run permutation test.
+
+        Args:
+        adata (anndata.AnnData): Annotated data matrix.
+        groupby (str): Key in adata.obs for grouping cells.
+        contrast (str): Name of the contrast group.
+        verbose (bool): Whether to print progress.
+
+        Returns:
+        pandas.DataFrame: Results of the permutation test, with columns:
+            - distance: distance between the contrast group and the group
+            - pvalue: p-value of the permutation test
+            - significant: whether the group is significantly different from the contrast group
+            - pvalue_adj: p-value after multiple testing correction
+            - significant_adj: whether the group is significantly different from the contrast group after multiple testing correction
+        """
+
+        if Distance(self.metric, self.obsm_key).metric_fct.accepts_precomputed:
+            # Much faster if the metric can be called on the precomputed
+            # distance matrix, but not all metrics can do that.
+            return self.test_precomputed(adata, groupby, contrast, verbose)
+        else:
+            return self.test_xy(adata, groupby, contrast, verbose)
+
+    def test_xy(self, adata: AnnData, groupby: str, contrast: str, verbose: bool = True) -> pd.DataFrame:
+        """Run permutation test for metrics that take x and y."""
+        distance = Distance(self.metric, self.obsm_key)
+        groups = adata.obs[groupby].unique()
+        if contrast not in groups:
+            raise ValueError(f"Contrast group {contrast} not found in {groupby} of adata.obs.")
+        fct = track if verbose else lambda x: x
+        emb = adata.obsm[self.obsm_key]
+        res = []
+
+        # Generate the null distribution
+        for _i in fct(range(self.n_perms)):
+            # per perturbation, shuffle with control and compute e-distance
+            df = pd.DataFrame(index=groups, columns=["distance"], dtype=float)
+            for group in groups:
+                if group == contrast:
+                    continue
+                # Shuffle the labels of the groups
+                mask = adata.obs[groupby].isin([group, contrast])
+                labels = adata.obs[groupby].values[mask]
+                shuffled_labels = np.random.permutation(labels)
+                idx = shuffled_labels == group
+
+                X = emb[mask][idx]  # shuffled group
+                Y = emb[mask][~idx]  # shuffled contrast
+                dist = distance(X, Y)
+
+                df.loc[group, "distance"] = dist
+            res.append(df.sort_index())
+
+        # Generate the empirical distribution
+        for group in groups:
+            if group == contrast:
+                continue
+            X = emb[adata.obs[groupby] == group]
+            Y = emb[adata.obs[groupby] == contrast]
+            df.loc[group, "distance"] = distance(X, Y)
+
+        # Evaluate the test
+        # count times shuffling resulted in larger distance
```

Comment:
Not sure what this comment is supposed to mean. What consequence does this have now?

Reply:
The comment explains how the samples from the null distribution are compared in a one-sided way against the empirical samples (I could also write it like this in the comment).